controller/api/destination/external-workload/endpoints_controller.go

```diff
@@ -0,0 +1,449 @@
+package externalworkload
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	ewv1alpha1 "github.com/linkerd/linkerd2/controller/gen/apis/externalworkload/v1alpha1"
+	"github.com/linkerd/linkerd2/controller/k8s"
+	logging "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+)
+
+const (
+	// Specifies capacity for updates buffer
+	updateQueueCapacity = 400
+
+	// Name of the lease resource the controllem will use
+	leaseName = "linkerd-destination-endpoint-write"
+
+	// Duration of the lease
+	leaseDuration = 30 * time.Second
+
+	// Deadline for the leader to refresh its lease. Core controllers have a
+	// deadline of 10 seconds.
+	leaseRenewDeadline = 10 * time.Second
+
+	// Duration a leader elector should wait in between action re-tries.
+	// Core controllers have a value of 2 seconds.
+	leaseRetryPeriod = 2 * time.Second
+)
+
+// EndpointsController reconciles service memberships for ExternalWorkload resources
+// by writing EndpointSlice objects for Services that select one or more
+// external endpoints.
+type EndpointsController struct {
+	k8sAPI   *k8s.API
+	log      *logging.Entry
+	updates  chan string
```

Comment:
I will spent a bit more time thinking about the best way to do that but it would be very useful to have: 

- metrics on current queue size
- metrics on avg time spent in the queue
- overflow counter

Reply:
Added some metrics for the queue. I re-shuffled bits of code around to capture some of the data (e.g. time spent in the queue). Let me know if this looks okay to you.

We're basically computing:
1. Histograms for queue latency (time spent in queue, the buckets were a bit arbitrary would love some feedback there)
2. Number of total elements enqueued
3. Queue length
4. Number of elements dropped from the queue (I thought this might make more sense than an overflow. For the translator we have an overflow counter per translator, once it overflows it kills the stream, we continue to work here so it felt better to capture how many updates we've shed).